### PR TITLE
Calypso contract - initial implementation

### PR DESCRIPTION
**Pull request by pierluca**
This does very little for now, it's roughly a copy of the value contract to deal with the advertisement of Secret Management Committees on the blockchain.

### CODE REVIEW

File: contracts/calypso/calypso.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	roster := step.Current.GetArg(RosterArg)
+	if len(roster) == 0 {
+		return xerrors.Errorf("'%s' not found in tx arg", RosterArg)
+	}
+
+	nodeList := strings.Split(string(roster), ",")
+	for _, r := range nodeList {
+		_, _, err := net.SplitHostPort(r)
+		if err != nil {
+			return xerrors.Errorf("invalid node '%s' in roster", r, err)
```
Comment (jbsv) [inline contracts/calypso/calypso.go:170]: need a %v for the error ?

File: contracts/calypso/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+var aKey = [32]byte{3}
+
+// miniController is a CLI initializer to register the value contract
+//
+// - implements node.Initializer
+type miniController struct {
+}
+
+// NewController creates a new minimal controller for the value contract.
+func NewController() node.Initializer {
```
Comment (jbsv) [inline contracts/calypso/controller/controller.go:22]: why node.Initializer and not calypso.Initializer ? I have to look in dela to understand where this comes from.

**Review by jbsv — APPROVED** [review]
LGTM so far. Minor things.

**Comment by jbsv** [comment]
it's weird. I verified the tests on MacOS and they all pass.

File: contracts/calypso/calypso.go (hunk 1 above)
Comment (pierluca) [reply contracts/calypso/calypso.go:170]: Yup. Oopsie.

File: contracts/calypso/calypso.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+)
+
+// Command defines a type of command for the value contract
+type Command string
+
+const (
+	// CmdAdvertiseSmc defines the command to advertise a SMC
+	CmdAdvertiseSmc Command = "ADVERTISE_SMC"
+
+	// CmdDeleteSmc defines a command to deleteSmc a SMC
```
Comment (jbsv) [inline contracts/calypso/calypso.go:56]:
```suggestion
	// CmdDeleteSmc defines a command to delete a SMC
```

File: contracts/calypso/calypso.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	CmdDeleteSmc Command = "DELETE_SMC"
+
+	// CmdListSmc defines a command to listSmc all SMCs (and not deleted) so far.
```
Comment (jbsv) [inline contracts/calypso/calypso.go:59]:
```suggestion
	// CmdListSmc defines a command to list a SMC (and not deleted) so far.
```

File: contracts/calypso/calypso.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+	return access.NewContractCreds(id, ContractName, credentialAllCommand)
+}
+
+// RegisterContract registers the value contract to the given execution service.
+func RegisterContract(exec *native.Service, c Contract) {
+	exec.Set(ContractName, c)
+}
+
+// Contract is a simple smart contract that allows one to handle the storage by
+// performing CRUD operations.
```
Comment (jbsv) [inline contracts/calypso/calypso.go:75]: not sure it is a CRUD operation as I don't see the update ?

File: contracts/calypso/calypso.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+// calypsoCommand implements the commands of the value contract
+//
+// - implements commands
+type calypsoCommand struct {
+	*Contract
+}
+
+// advertiseSmc implements commands. It performs the WRITE command
+func (c calypsoCommand) advertiseSmc(snap store.Snapshot, step execution.Step) error {
```
Comment (jbsv) [inline contracts/calypso/calypso.go:155]: does this CREATE and UPDATE ?

File: contracts/calypso/calypso_test.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	require.Nil(t, res)
+
+	_, found := contract.index[keyStr]
+	require.False(t, found)
+}
+
+func TestCommand_ListSmc(t *testing.T) {
+	contract := NewContract([]byte{}, fakeAccess{})
+
+	key1 := "key1"
```
Comment (jbsv) [inline contracts/calypso/calypso_test.go:121]:
```suggestion
	key1 := "roster1"
```

File: contracts/calypso/calypso_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+	key2 := "key2"
```
Comment (jbsv) [inline contracts/calypso/calypso_test.go:122]:
```suggestion
	key2 := "roster2"
```

File: contracts/calypso/calypso.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
+// Package calypso implements a Calypso contract that can advertise Secret
+// Management Committees and deal with secrets.
```
[the remark this answers is not in the document]
Comment (pierluca) [reply contracts/calypso/calypso.go:2]: multiple committees are supported :)

**Comment by pierluca** [comment]
Closing branch as this will go in HBT repository.

**Pull request closed without merging**